File: 0093-restore-ip-addresses/0093-restore-ip-addresses.py
class Solution:
  def restoreIpAddresses(self, s: str) -> List[str]:
    ans = []
    if len(s) not in range(4, 13):
      return ans
    if len(s) == 4:
      ans.append(".".join(list(s)))
      return ans
    
    def validOctet(s):
      if len(s) not in range(1, 4):
        return False
      if len(s) > 1 and s[0] == '0':
        return False
      if int(s) > 255:
        return False
      return True
    
    def dfs(s, octIdx=3, addr=''):
      if s == '':
        return
      if len(s) > (octIdx+1)*3:
        return
      if not octIdx and not validOctet(s):
        return
      if validOctet(s) and not octIdx:
        ans.append(s+addr)
        return
      
      for digits in range(1, min(len(s), 3)+1):
        oct = s[-digits:]
        if validOctet(oct):
          dfs(s[0:-digits], octIdx-1, f'.{oct}{addr}')
          
    dfs(s)
    return ans

File: 0093-restore-ip-addresses/0093-restore-ip-addresses.ts
function restoreIpAddresses(s: string): string[] {
  const ans: string[] = []
  if (s.length < 4 || s.length > 12) return ans
  if (s.length === 4) {
    ans.push(s.split("").join("."))
    return ans
  }
  
  function validOctet(s: string): boolean {
    if (s.length > 3 || s.length < 1) return false
    if (s.length > 1 && s[0] === '0') return false
    if (+s > 255) return false
    return true
  }

  function dfs(s: string, octIdx: number = 3, addr: string = '') {
    if (!s.length) return
    if (s.length > (octIdx + 1) * 3) return
    if (!octIdx && !validOctet(s)) return
    if (validOctet(s) && !octIdx) {
      ans.push(s + addr)
      return
    }
    for (let digits = 1; digits <= Math.min(s.length, 3); digits++) {
      const oct: string = s.slice(s.length - digits)
      if (validOctet(oct)) {
        dfs(s.slice(0, s.length - digits), octIdx - 1, `.${oct}${addr}`)
      }
    }
  }
  
  dfs(s)
  return ans
};